software/controller/lib/hal/i2c.h

```diff
@@ -293,14 +293,13 @@ class STM32Channel : public Channel {
   // handled in software.
   void Init(I2CReg *i2c, DMA::Base dma, Speed speed);
   // Interrupt handlers for DMA, which only makes sense on the STM32
-  void DMAIntHandler(DMA::Channel chan);
+  void DMAIntHandler(ExchangeDirection direction);
 
  private:
   I2CReg *i2c_{nullptr};
-  DMA::Base dma_;
-  /// \TODO: anyway to avoid keeping regiters here directly? Improve DMA abstraction?
-  volatile DmaReg::ChannelRegs *rx_channel_{nullptr};
-  volatile DmaReg::ChannelRegs *tx_channel_{nullptr};
+
+  std::optional<DMA::ChannelControl> rx_channel_control_{std::nullopt};
+  std::optional<DMA::ChannelControl> tx_channel_control_{std::nullopt};
```

Comment:
I came up with `rx_dma_` and `tx_dma_`during stepper refactoring instead of `rx_channel_control_` and `tx_channel_control_`, which I think are much clearer. Will rename those.

Reply:
> The only major problem that I see is some `SetupTx` and `SetupRx` calls don't appear to match up.

See above: in the context of DMA, Tx is Transfer, which applies whether you send or receive data, while in most serial communication (I2C, SPI and UART at least) Tx stands for Transmit, and Rx for Receive.
To clarify things, I'll rename `SetupTx` as `SetupTransfer`.

> Also, did you get to test the i2c classes for both reading and writing? Perhaps more nonvolatile memory tests?

Yes, I tested reading and writing to the EEPROM as well as checking the sensors (which use a2c, which in turn uses dma) still returned meaningful values (don't have a reference to compare to but it's similar to before) and the stepper movements.